### DISCUSSION TITLE
media_type not propagated to openapi.json

### DISCUSSION THREAD

**Discussion by tomasjerabek-com** [fastapi/fastapi#6964] [Questions] [answered]
### First check

* [ x ] I added a very descriptive title to this issue.
* [ x ] I used the GitHub search to find a similar issue and didn't find it.
* [ x ] I searched the FastAPI documentation, with the integrated search.
* [ x ] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ x ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ x ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
@app.get("/getmap/{filepath:path}", tags=["maps"])
def get_map(request: Request, filepath: str, auth: str = Header(None)):
    """Get map from path.<br><br>

    :param request: object<br>
    :param filepath: str<br>
    :param auth: str<br>
    :return:
    """
    logger.debug(f"{request.method} request to {request.url.path} endpoint from {request.client.host}.")
    if auth is None:
        raise credentials_exception
    validate_token(auth)
    if filepath.startswith("/"):
        filepath = filepath[1:]
    source_path = os.path.join(config.get("STORAGE", "PATH"), filepath)
    from os import path
    if not path.exists(source_path):
        raise HTTPException(status_code=404, detail=f"File {filepath} not found.")
    return FileResponse(source_path, media_type='application/zip')
```

### Description

Expected behaviour:
"Successful Response","content":{"**application/zip**"

Actual behaviour:
"Successful Response","content":{"application/json"

My workaround:
As media_type is ignored for openapi.json generation, I ended up using custom response class. 

```python
class ZipResponse(Response):
    media_type = "application/zip"

    def render(self, content: Any) -> bytes:
        return content
```
and using this as response class in method decorator:
```python
@app.get("/getmap/{filepath:path}", tags=["maps"], response_class=ZipResponse)
```

### Environment

* OS: [e.g. Linux / Windows / macOS]: macOS
* FastAPI Version [e.g. 0.3.0]: 0.61.2
* Python version: Python 3.9.1

**Comment by falkben** [comment, marked as answer]
Hi @tomasjerabek-com! Yup, that's exactly what the response class is there for. I'll try to explain below:

The default response class is `JSONResponse`. When creating the openapi schema fastapi does not inspect your entire function to figure out what responses you are actually returning. In order to change that, you need to explicitly mark your path operator, exactly as you have done so (or change it at the router or the app level if you want it applied to more than one). Hope that helps!

**Comment by tomasjerabek-com** [comment]
OK, thank you for your answer.
I was only questioning why media_type is ignored for openapi.json generation in this line: 
`return FileResponse(source_path, media_type='application/zip')`
As you say - fastapi inspects path operator (decorator) but does not care about return statement of that function? This is a bit confusing to me to know/remember what is being used for openapi.json and what is ignored.
Maybe this could be a feature request?

**Comment by Mause** [comment]
FastAPI doesn't do any analysis of function bodies to create the openapi schema.

Please refer to the documentation to see how to specify content/media types for routes: https://fastapi.tiangolo.com/advanced/additional-responses/

**Reply by benoit-arcsec** [reply]
Hello,

I am facing the same issue and this thread solved it. 
But still, why do I have to specify the `response_class` and it is not inferred from the function signature?

This works:

```python
@app.get("/files", response_class=ZipResponse)
def download() -> ZipResponse:
```

But this does not (it still display `application/json`in openapi.json):
```python
@app.get("/files")
def download() -> ZipResponse:
```

I expect that the media_type would be inferred from the function return type. Like the content is inferred when the return type is a pydantic model.

Note that it has the same behavior with `HTMLResponse` and `PlainTextResponse`.

### Environment

    OS: Linux
    FastAPI Version: 0.115.6
    Python version: Python 3.12.3

edit: code formatting

**Comment by tomasjerabek-com** [comment]
OK, thank you.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @tomasjerabek-com  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
